
<!doctype html>
    <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">	
    <link rel='shortcut icon' href='img/favicon.ico' type='image/x-icon'/ >
    <link type="text/css" rel="stylesheet" href="css/grid.css"  media="screen,projection" />
    <link type="text/css" rel="stylesheet" href="css/global.css"  media="screen,projection" />
    <link href="https://fonts.googleapis.com/css?family=Lato:300" rel="stylesheet"> 
    <title>ConGen2020 - Assembly Workshop</title>
</head>

<body>
    <div class="row" id="desktop_nav">
		<div class="col-2-24" id="margin"></div>
		<div class="col-4-24" id="nav_link_cell"><a class="nav_link" href="index.html">Intro</a></div>
		<!-- <div class="col-4-24" id="nav_link_cell"><a class="nav_link" href="start.html">Workshop</a></div> -->
		<div class="col-4-24 dropdown" id="nav_link_cell_active">
			<a href="start.html" class="nav_link" id="active">Workshop</a>
			<div class="dropdown_container desktop_drop">
				<ul class="dropdown_list">
					<li class="dropdown_list_item"><a href="start.html" id="dropdown_nav_link">Start</a></li>
					<li class="dropdown_list_item"><a href="reads.html" id="dropdown_nav_link">Reads</a></li>
					<li class="dropdown_list_item"><a href="#" id="dropdown_nav_link">Assembly</a></li>
					<li class="dropdown_list_item"><a href="mapping.html" id="dropdown_nav_link">Mapping</a></li>
					<li class="dropdown_list_item"><a href="iterative-mapping.html" id="dropdown_nav_link">Iterative mapping</a></li>
				</ul>
			</div>
		</div>
		<div class="col-4-24" id="nav_link_cell"><a class="nav_link" href="terms.html">Terminology</a></div>
		<div class="col-4-24" id="nav_link_cell"><a class="nav_link" href="programs.html">Programs</a></div>
		<div class="col-4-24" id="nav_link_cell"><a class="nav_link" href="links.html">Links</a></div>
		<div class="col-u-2-24" id="margin"></div>
	</div>

	<div class="row" id="mobile_nav">
		<div class="col-24-24 dropdown" id="nav_link_cell">
			<a href="#" class="nav_link"><img id="mobile_logo_nav" src="img/nav.png"></a>
			<div class="dropdown_container mobile_drop">
				<ul class="dropdown_list">
					<li class="dropdown_list_item"><a href="intro.html" id="mobile_nav_link">Intro</a></li>
					<li class="dropdown_list_item"><a href="start.html" id="mobile_nav_link">Workshop</a></li>
					<li class="dropdown_list_item"><a href="terms.html" id="mobile_nav_link">Terminology</a></li>
					<li class="dropdown_list_item"><a href="programs.html" id="mobile_nav_link">Programs</a></li>
					<li class="dropdown_list_item"><a href="links.html" id="mobile_nav_link">Links</a></li>
				</ul>
			</div>
		</div>
	</div>

    <a class="internal-link" name="top"></a>
    <div class="row" id="header">Genome assembly</div>

    <div class="row" id="body-row">
        <div class="col-3-24" id="side-nav-cont">
            <div id="side-nav">
                <span id="side-header">Page contents</span>
                <ul>
                    <li><a href="assembly.html#top">Intro</a></li>
                    <li><a href="assembly.html#spades">Short read assembly with SPAdes</a></li>
                    <li><a href="assembly.html#asm-qual">Assembly quality with N50</a></li>
                    <li><a href="assembly.html#busco">Assembly completeness with BUSCO</a></li>
                    <li><a href="assembly.html#index">Preparing assembly for analysis</a></li>
                </ul>
            </div>
        </div>

        <div class="col-21-24" id="main-cont">
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p><em>de novo</em> Genome assembly is the process by which overlapping sequence reads are pieced back together to form longer, contiguous sequences
                                for analysis:
                            </p>
                
                            <div class="row" id="img-row">
                                <div class="col-4-24" id="margin"></div>
                                <div class="col-16-24" id="img-col">
                                    <img id="res-img" src="img/assembly-flow-2.png">
                                    <center><span class="fig-caption">Figure 3.1: Genome assembly usually proceeds in steps that connect progressively longer segments of sequence.</span></center>
                                </div>
                                <div class="col-4-24" id="margin"></div>
                            </div>       
                            
                            <p>Many factors can play a role in the ultimate success of a genome assembly, including sequencing error rate, heterozygosity of the target genome, length of the target genome,
                                and number and length of reads sequenced.
                            </p>

                            <!-- <p>As we discussed in the lecture, there are a couple differences between assembling short-reads and assembling long-reads. First, the underlying representation of the contig assembly
                                graph. Short reads can employ a de Bruijn graph based approach that uses k-mers to further break down the reads to find overlaps. This has the advantage of preventing pairwise
                                comparisons of all reads, which is computationally intensive. The de Bruijn graph can also use Eulerian path algorithms, which are easier to solve, and employ error-correction
                                based on structures in the assembly graph (i.e. bubbles and tips).</p>

                            <p>While long read assemblies can be done using de Bruijn graph based approaches, -->

                            <p>Today, we'll be using both long and short reads to assemble the <em>D. pseudoobscura</em> genome and then using several metrics to assess the resulting assemblies.</p>

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>    

            <a class="internal-link" name="spades"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Assembly with short reads with <a href="https://github.com/ablab/spades/" target="_blank">SPAdes</a></div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>SPAdes is a de Bruijn graph based assembler that can handle many types of data, both short and long reads, and can scaffold contigs together using read pairs or pre-assembled contigs.
                                It is highly automated, and combines results from graphs built from multiple values of k. SPAdes, while relatively fast among genome assemblers, still takes a while to run, 
                                so today we'll be running one assembly of <em>D. pseudoobscura</em> on short reads only and compare it to a pre-made assembly that uses both long and short reads.

                            <p>Let's start by generating the short read assembly:

                            <center><pre class="cmd"><code>spades.py -1 /dpse-2-reads/illumina/chr2-1mil_1.fastq.gz -2 /dpse-2-reads/illumina/chr2-1mil_1.fastq.gz -o assemblies/spades-illumina -t 16 --isolate</code></pre></center>

                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">spades.py</td><td class="tcol-2">Call the main SPAdes interface script.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-1 /dpse-2-reads/illumina/chr2-1mil_1.fastq.gz</td><td class="tcol-2">The path to the first read-pair file.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-2 /dpse-2-reads/illumina/chr2-1mil_2.fastq.gz</td><td class="tcol-2">The path to the second read-pair file.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-o assemblies/spades-illumina/</td><td class="tcol-2">The path to the desired output directory.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-t 16</td><td class="tcol-2">The number of threads SPAdes can use.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">--isolate</td><td class="tcol-2">--isolate</td>
                                    </tr>
                                </table> 
                            </div>

                            <p>This will build the SPAdes assembly from the Illumina short reads. It should take about XX minutes to run.</p>

                            <p>SPAdes creates many output files within the specified output directory, including assemblies for multiple k-values. The final assembly will be constructed by comparing
                                portions of each of the k-mer graphs and selecting the most resolved regions
                            </p>

                            <p>The final assembly will be in the file <code class="inline">scaffolds.fasta</code>

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="asm-qual"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Assessing assembly quality</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>One of the main goals for a genome assembly program is to produce assembly scaffolds approaching chromosome length. This is difficult because repeats can often form 
                                unresolvable portions of the underlying assembly graph that prevent the extension of a sequence. This means genome assemblies can be highly fragmented.</p>

                            <p>Here are some important statistics to consider when assessing a fragmented assembly:</p>

                            <div class="table-cont">
                                <table class="norm-table">
                                    <thead><th class="tcol-1">Statistic</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">Total assembly size</code></td><td class="tcol-2">The sum of the length of all contigs. 
                                            If you have an expected genome size from a closely related organism or other analysis, this can help determine how much of the genome you have assembled.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">N50</code></td><td class="tcol-2">The length, N, of the scaffold/contig such that 50% ofo the 
                                            assembly is contained in scaffolds of length N and longer. In other words, at least half of the nucleotides in the assembly belongs to contigs with the 
                                            N50 length or longer.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">L90</code></td><td class="tcol-2">The number of scaffolds that make up 90% of the assembly. Ideally close to the
                                            number of chromosomes in the sequenced organism.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">Number of Ns</code></td><td class="tcol-2">Bases that can't be confidently called, especially near contig junctions within
                                            scaffolds, are represented by Ns in the FASTA assembly file. Counting the total number of Ns can give you an idea of how well contigs were joined together
                                            into scaffolds.</td>
                                    </tr>
                                </table>
                            </div>

                            <p><code class="inline">N50</code> can be a little tricky to understand, luckily the Molecular Ecologist blog
                                has a <a href="https://www.molecularecologist.com/2017/03/whats-n50/" target="_blank">simple explainer post</a>, that has the following figure:

                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="res-img" src="img/n50.png">
                                    <center><span class="fig-caption">Figure 3.2: Scaffold N50. Source: 
                                        <a href="https://www.molecularecologist.com/2017/03/whats-n50/" target="_blank">Molecular Ecologist</a></span></center>
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>

                            <p><code class="inline">L90</code> is similar, but using number of scaffolds instead of scaffold length.</p>

                            <p>These statistics are relatively easy to calculate, but there are also tools that can summarize these and more for a given assembly.
                                <a href="https://doi.org/10.1186/2047-217X-2-10" target="_blank">Assemblathon</a> was a competition to assess genome assemblers in 2013. They have helpfully made available
                                some of the scripts they use to quantify final genome assemblies in their <a href="https://doi.org/10.1186/2047-217X-2-10" target="_blank">github repository</a>. We will 
                                use one of their scripts to gather some basic statistics about the genome assembly we just produced:
                            </p>

                            <center><pre class="cmd"><code>perl /usr/bin/assemblathon_stats.pl assemblies/illumina-only-spades/scaffolds.fasta > assemblies/illumina-assemblathon.txt</code></pre></center>

                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">perl</td><td class="tcol-2">Invoke <code class="inline">perl</code> to run a script.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">/usr/bin/assemblathon_stats.pl</td><td class="tcol-2">The path to the <code class="inline">perl</code> script to run.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/illumina-only-spades/scaffolds.fasta</td><td class="tcol-2">The input FASTA file containing a genome assembly.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">&gt;</td><td class="tcol-2">This redirects the output from the command specified before it into the file specified after it.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-t 16</td><td class="tcol-2">The number of threads SPAdes can use.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/illumina-assemblathon.txt</td><td class="tcol-2">The desired output file.</td>
                                    </tr>
                                </table> 
                            </div>                        

                            <p>This should produce the file <code class="inline">assemblies/illumina-assemblathon.txt</code> which contains several statistics about the assembly</p>

                            <p>Let's take a look at this file:</p>
                            
                            <center><pre class="cmd"><code>less -S assemblies/illumina-assemblathon.txt</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">less</td><td class="tcol-2">A Linux text file viewer 
                                            (use <code class="inline">&lt;up arrow&gt;</code> and <code class="inline">&lt;down arrow&gt;</code> to scroll up and down by one line 
                                            or <code class="inline">&lt;space&gt;</code> and <code class="inline">b</code> 
                                            to scroll up and down by one page, resepectively. Type <code class="inline">q</code> to exit)</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-S</td><td class="tcol-2">Turn off line-wrapping within <code class="inline">less</code> 
                                            (use the <code class="inline">&lt;left arrow&gt;</code> and <code class="inline">&lt;right arrow&gt;</code> to scroll left and right).</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/illumina-assemblathon.txt</td><td class="tcol-2">The path to the file you want to view</td>
                                    </tr>
                                </table>
                            </div>

                            <p>This should display a nice table of statstics:</p>

                            <pre class="text"><code>The table.</code></pre>

                            <p>We know that <em>D. pseudoobscura</em> has a genome size of roughly 150Mb in 6 chromosomes. So the assembly size of <code class="inline">159563605</code>
                                is about right, but we've only been able to resolve <code class="inline">133829</code> scaffolds. No where near the 6 chromosomes we would ideally want.
                                Such a fragmented assembly would make annotation difficult with genes likely being split between scaffolds.
                            </p>

                            <p>What about some other assemblies? We've provided several more assemblies of <em>D. pseudoobscura</em> in the <code class="list">expected-outputs/assemblies/</code> folder,
                                including a SPAdes assembly with long reads (PacBio + Nanopore), a <a href="https://github.com/fenderglass/Flye/" target="_blank">Flye</a>
                                assembly with just PacBio reads and a Flye assembly with just Nanopore reads. The assemblathon output files are there as well, but the statistics we've talked about
                                are summarized in the following table:
                            </p>

                            <div class="table-cont">
                                <table class="norm-table">
                                    <thead>
                                        <th>Statistic</th><th><a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_009870125.1/" target="_blank">SRA assembly</a> (PacBio/Canu)</th>
                                        <th>SPAdes (Illumina only)</th>
                                        <th>SPAdes (PacBio+Nanopore)</th>
                                        <th>Flye (PacBio)</th>
                                        <th>Flye (Nanopore)</th>
                                    </thead>
                                    <tr>
                                        <td>Number of scaffolds</td>
                                        <td>70</td>
                                        <td>133829</td>
                                        <td>122980</td>
                                        <td>239</td>
                                        <td>1321</td>
                                    </tr>
                                    <tr>
                                        <td>Assembly size</td>
                                        <td>163282969</td>
                                        <td>159563605</td>
                                        <td>159513481</td>
                                        <td>158355727</td>
                                        <td>151695649</td>
                                    </tr>
                                    <tr>
                                        <td>Scaffold N50</td>
                                        <td>32422566</td>
                                        <td>65232</td>
                                        <td>110786</td>
                                        <td>17300572</td>
                                        <td>5886457</td>
                                    </tr>
                                    <tr>
                                        <td>L90</td>
                                        <td>NA</td>
                                        <td>12954</td>
                                        <td>7018</td>
                                        <td>15</td>
                                        <td>70</td>
                                    </tr>
                                    <tr>
                                        <td>% Ns</td>
                                        <td>NA</td>
                                        <td>0.12</td>
                                        <td>0.05</td>
                                        <td>0.00</td>
                                        <td>0.00</td>
                                    </tr>
                                </table>
                            </div>
                        
                            <p>Clearly the long read assemblies with the long read assembler (Flye) have higher N50s and fewer scaffolds than the SPAdes assemblies. Note that this doesn't mean SPAdes
                                is "worse" than Flye, it just means that for this type of data Flye is producing more contiguous assemblies. Also note that none of these assemblies alone produce scaffolds
                                approaching the SRA assembly for <em>D. pseudoobscura</em> (PacBio/Canu), which has only 70 scaffolds, most assigned to one of the 6 chromosomes.
                            </p>

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="busco"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Annotation quality with <a href="https://busco.ezlab.org/" target="_blank">BUSCO</a></div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>BUSCO, short for Benchmarking sets of Universal Single-Copy Orthologs, can help to assess the completeness of your genome with respect to gene content.
                                It does so by comparing gene sequences of well annotated species closely related to your species of interest to determine the number of Complete, Duplicated,
                                Fragmented, and Missing genes present in an assembly among a set of universal orthologs.
                            </p>

                            <p>You'll learn more about BUSCO in the next session, but for now we'll just take a look at a sample output from our SPAdes Illumina assembly.</p>

                            <pre class="text"><code>The table.</code></pre>

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="index"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Preparing your assembly for analysis</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>Now that we have a FASTA file that contains our genome assembly, and we know roughly how complete it is from a technical (N50) and biological (BUSCO)
                                perspective, we're ready to use it for analysis. Many downstream analyses require an assembly FASTA to be <b>indexed</b>. There are several types of 
                                indices and I usually make all of them for any assembly I work with. These indices are made with <code class="inline>samtools faidx</code>,
                                <code class="inline>bwa index</code>, and <code class="inline">picard CreateSequenceDictionary</code>.
                            </p>

                            <p>All of these are simple to generate. Today we'll just generate the <code class="inline">samtools faidx</code> for our SPAdes Illumina assembly:</p>

                            <center><pre class="cmd"><code>samtools faidx assemblies/spades-illumina/scaffolds.fasta</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">faidx</td><td class="tcol-2">Call the <code class="inline">faidx</code> sub-program within <code class="inline">samtools</code></td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/spades-illumina/scaffolds.fasta</td><td class="tcol-2">The path to the FASTA file to index</td>
                                    </tr>
                                </table>
                            </div>

                            <p>This should run fairly quickly and will produce the file <code class="inline">assemblies/spades-illumina/scaffolds.fasta.fai</code> file that we can look at</p>

                            <center><pre class="cmd"><code>less -S assemblies/spades-illumina/scaffolds.fasta.fai</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">less</td><td class="tcol-2">A Linux text file viewer 
                                            (use <code class="inline">&lt;up arrow&gt;</code> and <code class="inline">&lt;down arrow&gt;</code> to scroll up and down by one line 
                                            or <code class="inline">&lt;space&gt;</code> and <code class="inline">b</code> 
                                            to scroll up and down by one page, resepectively. Type <code class="inline">q</code> to exit)</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-S</td><td class="tcol-2">Turn off line-wrapping within <code class="inline">less</code> 
                                            (use the <code class="inline">&lt;left arrow&gt;</code> and <code class="inline">&lt;right arrow&gt;</code> to scroll left and right).</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/spades-illumina/scaffolds.fasta.fai</td><td class="tcol-2">The path to the file you want to view</td>
                                    </tr>
                                </table>
                            </div>

                            <pre class="text"><code>The table.</code></pre>

                            <p><code class="inline">.fai</code> files are pretty straightforward. They contain tab-delimited columns with information for one scaffold on each row. Column 1
                                indicates the scaffold name, Column 2 the length, Column 3 a byte index, Column 4 the number of bases per line, and Column 5 the number of bytes per line
                            </p>

                            <p>Other indices that we won't generate directly today include the BWA index files, which make it easier to map reads to an assembly:

                            <center><pre class="cmd"><code>bwa index assemblies/spades-illumina/scaffolds.fasta</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">bwa</td><td class="tcol-2">Call the <code class="inline">bwa</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">index</td><td class="tcol-2">Call the <code class="inline">index</code> sub-program within <code class="inline">bwa</code></td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">assemblies/spades-illumina/scaffolds.fasta</td><td class="tcol-2">The path to the FASTA file to index</td>
                                    </tr>
                                </table>
                            </div>                            

                            <p>And the Picard Dictionary file, which used by variant callers such as GATK and other programs:

                            <center><pre class="cmd"><code>Picard CreateSequenceDictionary I=assemblies/spades-illumina/scaffolds.fasta O=assemblies/spades-illumina/scaffolds.dict</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">picard</td><td class="tcol-2">Call the <code class="inline">picard</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">CreateSequenceDictionary</td><td class="tcol-2">Call the <code class="inline">CreateSequenceDictionary</code> 
                                            sub-program within <code class="inline">picard</code></td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">I=assemblies/spades-illumina/scaffolds.fasta</td><td class="tcol-2">The path to the FASTA file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">O=assemblies/spades-illumina/scaffolds.dict</td><td class="tcol-2">The path to the desired output file</td>
                                    </tr>
                                </table>
                            </div>

                            <p>After you've generated and assessed the quality of your assembly, the next step is usually annotation. This usually means figuring out where repeats
                                are in the assembly and classifying their types, and figuring out where the genes in the assembly are. There are many ways to do this, enough to fill
                                up a separate workshop, so we won't be covering annotation today.</p>

                            <p>Next we will talk about an alternate to de novo genome assembly: reference guided assembly with <a href="mapping.html">read mapping</a>...

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>


        </div>
    </div>

    <div class="row" id="btm-nav">
        <div class="col-3-24" id="nav-bnt-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="reads.html">&lt;&nbsp;Previous</a>    
            </div>
        </div>
        <div class="col-6-24" id="nav-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="mapping.html">Next&nbsp;&gt;</a>
            </div>
        </div>
        <div class="col-3-24" id="nav-btn-margin"></div>
    </div>

    <div class="row" id="footer">
		<div class="col-1">
			<div id="footer_text">
				<center>Site designed and maintained by <a href="https://gwct.github.io" target="_blank">Gregg Thomas</a> | Page built: 09/01/2020 23:49:47 MDT</center>
			</div>
		</div>
	</div>
</body>
