
<!doctype html>
    <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">	
    <link rel='shortcut icon' href='img/favicon.ico' type='image/x-icon'/ >
    <link type="text/css" rel="stylesheet" href="css/grid.css"  media="screen,projection" />
    <link type="text/css" rel="stylesheet" href="css/global.css"  media="screen,projection" />
    <link href="https://fonts.googleapis.com/css?family=Lato:300" rel="stylesheet"> 
    <title>UVU Genomics</title>
</head>

<body>
    <div class="row" id="desktop_nav">
		<div class="col-3-24" id="margin"></div>
		<div class="col-3-24" id="nav_link_cell"><a class="nav_link" href="index.html">Home</a></div>
		<div class="col-3-24" id="nav_link_cell"><a class="nav_link" href="syllabus.html">Syllabus</a></div>
		<div class="col-3-24 dropdown" id="nav_link_cell">
			<a href="lab01-1.html" class="nav_link">Lab Assignments</a>
			<div class="dropdown_container desktop_drop">
				<ul class="dropdown_list">
					<!-- <li class="dropdown_list_item"><a href="#" id="dropdown_nav_link">Review</a></li> -->
					<li class="dropdown_list_item"><a href="lab01-1.html" id="dropdown_nav_link">Lab 01</a></li>
					<li class="dropdown_list_item"><a href="assembly.html" id="dropdown_nav_link">Lab 02</a></li>
					<li class="dropdown_list_item"><a href="#" id="dropdown_nav_link">Lab 03</a></li>
					<!-- <li class="dropdown_list_item"><a href="iterative-mapping.html" id="dropdown_nav_link">Iterative mapping</a></li> -->
				</ul>
			</div>
		</div>
		<div class="col-3-24" id="nav_link_cell"><a class="nav_link" href="terms.html">Terminology</a></div>
		<div class="col-3-24" id="nav_link_cell"><a class="nav_link" href="programs.html">Programs</a></div>
		<div class="col-3-24" id="nav_link_cell"><a class="nav_link" href="links.html">Links</a></div>
		<div class="col-3-24" id="margin"></div>
	</div>

	<div class="row" id="mobile_nav">
		<div class="col-24-24 dropdown" id="nav_link_cell">
			<a href="#" class="nav_link"><img id="mobile_logo_nav" src="img/nav.png"></a>
			<div class="dropdown_container mobile_drop">
				<ul class="dropdown_list">
					<li class="dropdown_list_item"><a href="index.html" id="mobile_nav_link">Home</a></li>
					<li class="dropdown_list_item"><a href="syllabus.html" id="mobile_nav_link">Syllabus</a></li>
					<li class="dropdown_list_item"><a href="lab01-1.html" id="mobile_nav_link">Lab Assignments</a></li>
					<li class="dropdown_list_item"><a href="terms.html" id="mobile_nav_link">Terminology</a></li>
					<li class="dropdown_list_item"><a href="programs.html" id="mobile_nav_link">Programs</a></li>
					<li class="dropdown_list_item"><a href="links.html" id="mobile_nav_link">Links</a></li>
				</ul>
			</div>
		</div>
	</div>

        <a class="internal-link" name="top"></a>
    <div class="row" id="header">Read Mapping</div>

    <div class="row" id="body-row">
        <div class="col-3-24" id="side-nav-cont">
            <div id="side-nav">
                <span id="side-header">Page contents</span>
                <ul>
                    <li><a href="mapping.html#top">Intro</a></li>
                    <li><a href="mapping.html#bwa">Mapping reads with BWA</a></li>
                    <li><a href="mapping.html#sam">SAM/BAM format</a></li>
                    <li><a href="mapping.html#mkdup">Marking duplicate reads</a></li>
                    <li><a href="mapping.html#depth">Read depth</a></li>
                </ul>
            </div>
        </div>

        <div class="col-21-24" id="main-cont">
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>Now that we're familiar with de novo assembly, let's look at another strategy for genome assembly: reference-guided assembly by read mapping. Using this strategy, the
                                raw reads aren't put back together, but rather aligned to a reference genome:
                            </p>
                
                            <div class="row" id="img-row">
                                <div class="col-4-24" id="margin"></div>
                                <div class="col-16-24" id="img-col">
                                    <img id="res-img" src="img/ref-guided-assembly-1.png">
                                    <center><span class="fig-caption">Figure 4.1: Reference guided assembly via read-mapping.</span></center>
                                </div>
                                <div class="col-4-24" id="margin"></div>
                            </div>       
                            
                            <p>Reference-guided assembly with read mapping has several advantages, including being fast and obviating the need for de novo genome annotation (as long as the
                                reference genome is annotated). On the other hand, read mapping limits the type of mutations that can be inferred. For instance, while structural variation can
                                be indirectly inferred based on relative read-depth, large genome rearrangements between the reference and the sample cannot be captured simply by read mapping.
                                In other words, there's a lot about genome evolution that cannot be answered by mapping reads. Nevertheless, its utility and ease makes it a go-to strategy for
                                systems with good reference genomes.
                            </p>
                
                            <p>We'll be using BWA to map some short reads from D. pseudoobscura chromosome 2 to the homologous chromosome 3 in D. melanogaster and then assessing the read mapping.</p>
                
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="bwa"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Mapping reads with <a href="https://github.com/lh3/bwa" target="_blank">BWA</a> and <a href="http://www.htslib.org/" target="_blank">samtools</a></div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>BWA is a short-read alignment program and is very straightforward. Simply tell it the reads you want to map and the reference genome and it will take care of the rest:</p>

                            <center><pre class="cmd"><code>bwa mem -t 4 dmel-3R-reference/Drosophila_melanogaster.BDGP6.28.dna.chromosome.3R.fa dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz dpse-chr2-reads/illumina/pse-chr2_2.fastq.gz | samtools sort > mapped-reads/dpse2-to-dmel3R.bam</code></pre></center>
            
                            <p>You'll notice that in this command we're actually taking the output from BWA and piping it directly into <code class="inline">samtools</code> to sort the resulting read
                                mappings by coordinate. This sorting is required by many programs that process mapped reads.
                            </p>
            
                            <p>Here's a table that breaks down this command:</p>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">bwa</td><td class="tcol-2">Call the BWA program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">mem</td><td class="tcol-2">Use the <code class="inline">mem</code> alignment algorithm implemented in BWA</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-t 4</td><td class="tcol-2">Tells BWA to use 4 threads to map reads</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dmel-3R-reference/Drosophila_melanogaster.BDGP6.28.dna.chromosome.3R.fa</td><td class="tcol-2">The path to the reference genome file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz</td><td class="tcol-2">The path to the FASTQ file containin the first pairs of reads</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse-chr2-reads/illumina/pse-chr2_2.fastq.gz</td><td class="tcol-2">The path to the FASTQ file containin the second pairs of reads</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">|</td><td class="tcol-2">This pipe means that the output from the command specified before it will be used as input to the command specified after it</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">sort</td><td class="tcol-2">Call the sorting algorithm implemented <code class="inline">samtools</code></td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">&gt;</td><td class="tcol-2">This redirects the output from the command specified before it into the file specified after it</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">mapped-reads/dpse2-to-dmel3R.bam</td><td class="tcol-2">The output file for the mapped and sorted reads</td>
                                    </tr>
                                </table> 
                            </div>

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="sam"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Mapped reads in SAM/BAM format</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>After this runs, we should get the file <code class="inline">mapped-reads/dpse2-to-dmel3R.bam</code> which contains the mapped reads in BAM format. SAM format is a text format
                                that contains one mapped read per line, along with information about the mapping and BAM is just the compressed version of SAM format.
                            </p>
            
                            <p><code class="inline">samtools view</code> is a helpful tool that let's us retrieve lines from these files and convert between formats. SAM files also contain headers that
                                tell us about how the file was generated and the reads that make up the file. Let's use samtools view to look at the header of our newly mapped reads.
                            </p>
            
                            <p>First, let's change directories to the folder that contains the BAM file so we don't have to keep typing that in our commands:</p>
            
                            <center><pre class="cmd"><code>cd mapped-reads</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">cd</td><td class="tcol-2">The Linux change directory command</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">mapped-reads</td><td class="tcol-2">The path to the directory we want to change to</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>Now let's look at the header of our BAM file:</p>
            
                            <center><pre class="cmd"><code>samtools view -H dpse2-to-dmel3R.bam</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">view</td><td class="tcol-2">Use the <code class="inline">view</code> sub-program implemented within samtools</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-H</td><td class="tcol-2">Only output the header information of the input file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R.bam</td><td class="tcol-2">The input file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">|</td><td class="tcol-2">This pipe means that the output from the command specified before it will be used as input to the command specified after it</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code>head</code></td><td class="tcol-2">The Linux <code class="inline">head</code> program which returns only the first few lines of an input file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-n 3</td><td class="tcol-2">Tells <code class="inline">head</code> to return only 3 lines of output</td>
                                    </tr>                        
                                </table>
                            </div>
            
                            <p>The output should look something like this:</p>
            
                            <pre class="text"><code>@HD     VN:1.6  SO:coordinate
@SQ     SN:3R   LN:32079331
@PG     ID:bwa  PN:bwa  VN:0.7.17-r1198-dirty   CL:bwa mem -t 4 dmel-3R-reference/Drosophila_melanogaster.BDGP6.28.dna.chromosome.3R.fa dpse-2-reads/illumina/pse-chr2_1.fastq.gz dpse-2-reads/illumina/pse-chr2_2.fastq.gz
@PG     ID:samtools     PN:samtools     PP:bwa  VN:1.10 CL:samtools sort
@PG     ID:samtools.1   PN:samtools     PP:samtools     VN:1.10 CL:samtools view -H dpse2-to-dmel3R.bam</code></pre>

                            <p>The first line starting with <code class="inline">@HD</code> tells us the version of the SAM format that this file follows along with the sort order (<code class="inline">SO</code> of the file.
                                In this case, since we ran <code class="inline">samtools sort</code>, the sort order is coordinate.
                            </p>
            
                            <p>Lines starting with <code class="inline">@SQ</code> tell us about the sequence names and lengths of the reference genome. These would be the sequence headers from the reference FASTA
                            file.</p>
            
                            <p>Finally, <code class="inline">@PG</code> tells us about the programs that were used to generate this file. These should look familiar!</p>
            
                            <p>Without the <code class="inline">-H</code> option, we can view the actual entries in the BAM file:</p>
            
                            <center><pre class="cmd"><code>samtools view dpse2-to-dmel3R.bam | head -n 3</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code>view</code></td><td class="tcol-2">Use the <code class="inline">view</code> sub-program implemented within samtools</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R.bam</td><td class="tcol-2">The input file</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>Now you should see some actual mapped reads!</p>
                            <pre class="text"><code>SRR11230564.10809749    163     3R      33475   0       128S23M =       33475   23      GTAGACCCGACCGTTTTCCGGCCCAGATGACCTCCCCGATGCATAATATGTGCACCTGTCAACCAGAAGGTTCGTTTCTTAGGATTTCTATACTACAGTGCTGAGTAATTTTTCTCGCACTGTTTTTGGTAAAATAAAAAAAAAAAAAAAA    &lt;AAAFJFFJJJFJFFJJJJJJJJAFJJFFJFJFJJJJJJJJJJFFJJJJJJJJJJJAJFAJFJJJJFJJJFFFF7-7--7AA7JA&lt;&lt;AFJJJFJJJJFAFFFJJFAJJJAFFFJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ-    NM:i:0  MD:Z:23 MC:Z:128S23M    AS:i:23 XS:i:23
SRR11230564.10809749    83      3R      33475   0       128S23M =       33475   -23     GTAGACCCGACCGTTTTCCGGCCCAGATGACCTCCCCGATGCATAATATGTGCACCTGTCAACCAGAAGGTTCGTTTCTTAGGATTTCTATACTACAGTGCTGAGTAATTTTTCTCGCACTGTTTTTGGTAAAATAAAAAAAAAAAAAAAC    AF77)JJJJJJJJFJJAJJJJJJJJJFJJFJF&lt;FJJJJJJJJJJJJFJJFJJFFJJJFJJJJJJJJFJJJJJJJJJJJJJJJFJJJJFJJJJJJJFJFJJJJJJJJJJJJJJJ&lt;JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA    NM:i:1  MD:Z:22A0       MC:Z:128S23M    AS:i:22 XS:i:22
SRR11230564.16812502    99      3R      35215   0       55S20M76S       =       35233   38      GAGAGACCAGGCCCATAAAAACCTAAAGACGAAACTCGGCCAGTACGGGCGGTAAGAGTACGAGTACGGGACCGATTATTACACTTAAATTGTAACCAATTTAACGGCACGTAAAAACTAAAAGCAGTCTCGGTGAGATCGGAAGAGCGTC    AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFJJJJJAA    NM:i:0  MD:Z:20 MC:Z:71S20M60S  AS:i:20 XS:i:20</code></pre>
            
                            <p>This is a tab delimited output with one read mapping per line. Some of the columns should look familiar, like the read name (column 1), sequence (column 10),
                                and quality scores (column 11) from the FASTQ file. The other columns are described on the <a href="http://bio-bwa.sourceforge.net/bwa.shtml#4" target="_blank">BWA website</a>
                                or in the table in <a href="https://genomics.sschmeier.com/ngs-mapping/index.html#table-sam" target="_blank">this helpful walkthrough</a> that I've copied below:
                            </p>
            
                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="res-img" src="img/sam-cols.png">
                                    <center><span class="fig-caption">Figure 4.2: The column definitions of SAM/BAM formatted files. <a href="https://genomics.sschmeier.com/ngs-mapping/index.html#table-sam" target="_blank">Source</a>.</span></center>
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>                    

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="mkdup"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Marking duplicate reads in SAM/BAM format</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>Duplicate reads can lead to spurious results for analyses that rely on depth of coverage, such as inferences of copy number variation. Depending on the sequencing
                                machine used, the cause of duplicate reads can vary. For Illumina machines, duplicates can be caused by uneven PCR amplification (PCR duplicates) and a single
                                spot on the flow cell being mistakenly identified as two separate spots (optical duplicates).
                            </p>
            
                            <p>The best practice is to mark these duplicate reads in your final SAM/BAM file so that programs in later analyses can deal with them appropriately. To do this, 
                                we use the program Picard.
                            </p>
            
                            <center><pre class="cmd"><code>picard MarkDuplicates I=dpse2-to-dmel3R.bam O=dpse2-to-dmel3R.mkdup.bam VALIDATION_STRINGENCY=LENIENT M=dpse2-to-dmel3R-dupmets.txt CREATE_INDEX=true</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">picard</td><td class="tcol-2">Call the picard program</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">MarkDuplicates</td><td class="tcol-2">Use the MarkDuplicates sub-program implemented within picard</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">I=dpse2-to-dmel3R.bam</td><td class="tcol-2">The input file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">O=dpse2-to-dmel3R.mkdup.bam</td><td class="tcol-2">The desired output file</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">VALIDATION_STRINGENCY=LENIENT</td><td class="tcol-2">Tells picard not to exit if it encounters slightly different SAM formatting than expected</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">M=dpse2-to-dmel3R-dupmets.txt</td><td class="tcol-2">The name for a file to output some metrics</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">CREATE_INDEX=true</td><td class="tcol-2">Tells picard to also create and index file for the output BAM file. 
                                            These indices are necessary for other programs to use the file.</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>This should create a new BAM file, <code class="inline">dpse2-to-dmel3R.mkdup.bam</code>, as well as an index for this BAM file and a metrics file. In this file,
                                reads that have been determined to be duplicates will have <code class="inline">1024</code> added to their <code class="inline">FLAG</code> (column 2) in the
                                output BAM/SAM file. See <a href="https://broadinstitute.github.io/picard/explain-flags.html" target="_blank">this page</a> for more info about SAM FLAGS.
                            </p>
            
                            <p>We can get a summary of how many reads were marked as duplicates from the metrics file:</p>
            
                            <center><pre class="cmd"><code>grep -a2 "## METRICS CLASS" dpse2-to-dmel3R-dupmets.txt</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">grep</td><td class="tcol-2">The Linux pattern search command. Given a string and a file, <code class="inline">grep</code>
                                            searches the file for the string and prints lines that contain matches to the screen. Can be used with regular expression pattern matching.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-a2</td><td class="tcol-2">Tells <code class="inline">grep</code> to also print the two lines after the match is found.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">"## METRICS CLASS"</td><td class="tcol-2">The pattern we want to search for.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R-dupmets.txt</td><td class="tcol-2">The file in which we want to search for the specified pattern.</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>If the metrics file has been created succesfully, this should produce the following output:</p>
            
                            <pre class="text"><code>## METRICS CLASS        picard.sam.DuplicationMetrics
LIBRARY UNPAIRED_READS_EXAMINED READ_PAIRS_EXAMINED     SECONDARY_OR_SUPPLEMENTARY_RDS  UNMAPPED_READS  UNPAIRED_READ_DUPLICATES        READ_PAIR_DUPLICATES    READ_PAIR_OPTICAL_DUPLICATES       PERCENT_DUPLICATION     ESTIMATED_LIBRARY_SIZE
Unknown Library 844     200412  2766    673346  134     621     0       0.003426        32272106</code></pre>
            
                            <p>Based on the column headers, columns 6, 7, and 8 tell us how many duplicate reads were identified. In this case, 134 unpaired read duplicates, 
                                621 paired read duplicats, and 0 optical duplicates were found, for a total of 134 + 621x2 = 1376 duplicate reads.</p>
            
                            <p>If we wanted to, we could look at these reads using <code class="inline">samtools view</code> and the flag 1024:</p>
            
                            <center><pre class="cmd"><code>samtools view -f 1024 dpse2-to-dmel3R.mkdup.bam</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">view</td><td class="tcol-2">Use the <code class="inline">view</code> sub-program within <code class="inline">samtools</code>.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-f 1024</td><td class="tcol-2">Only look for lines that have this included in their <code class="inline">FLAG</code> column
                                            (column 2). 1024 is the MarkDuplicates flag.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R.mkdup.bam</td><td class="tcol-2">The input BAM/SAM file to view.</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>This should print all 1376 duplicate reads to the screen.</p>
            
                            <p>From here, the mapped reads can be used to calculate genotype likelihoods at each position to infer where the reference genome and the mapped genome differ. These variants can be used
                                in downstream analyses to assess divergence, phylogenetic relationships, selection, or any other comparative analysis you can think of.
                            </p>
            
                            <div class="row" id="img-row">
                                <div class="col-4-24" id="margin"></div>
                                <div class="col-16-24" id="img-col">
                                    <img id="res-img" src="img/ref-guided-assembly-2.png">
                                    <center><span class="fig-caption">Figure 4.3: Variant calling from mapped reads. Green sites are true variants while dark red sites are sequencing errors.</span></center>
                                </div>
                                <div class="col-4-24" id="margin"></div>
                            </div>
                            
                            <p>We won't cover this in detail today, but we've provided a list of several variant calling programs on the <a href="programs.html" target="_blank">Programs</a> page.</p>                        

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="depth"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Assessing mapped reads with read depth</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">

                            <p>One of the most common questions you'll hear when mapping reads to a reference (or even during de novo assembly) is what is the average read depth?
                                Read depth is simply defined as how many reads align to a given position in the reference genome and is related to the term coverage. However, 
                                when referring to coverage one is often asking how many reads it would take to cover X number of positions in the genome. With so many reads
                                being produced by modern sequencing machines, the two terms have become interchangeable because it is assumed all positions in the genome will
                                be covered by reads.
                            </p>
            
                            <p>One way for estimating coverage is to use the Lander/Waterman (<a href="https://doi.org/10.1016/0888-7543(88)90007-9" target="_blank">1988</a>) equation:</p>
            
                            <p><span class="math display" id="eqn">\[ Coverage = \frac{LN}{G} \]</span></p>
            
                            <div class="table-cont">
                                <table class="norm-table">
                                    <thead><th class="tcol-1">Term</th><th class="tcol-2">Definition</th></thead>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">Coverage</code></td><td class="tcol-2">The number of times a position in the genome has been sequenced. The number of reads "covering" a position.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">L</code></td><td class="tcol-2">Read length.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">N</code></td><td class="tcol-2">Number of reads.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1"><code class="inline">G</code></td><td class="tcol-2">Haploid genome length</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>We can use this equation in two ways:</p>
            
                            <ol>
                                <li>Estimate read depth/coverage by solving as is.</li>
                                <li>Estimate the number of reads needed to obtain a certain amount of coverage by solving for N.</li>
                            </ol>
            
                            <p>In our case, for <em>D. pseudoobscura</em> chromosome 2, we know <code class="inline">L</code>, <code class="inline">N</code>, and <code class="inline">G</code> from
                                our <code class="inline">fastqc</code> analysis and genome assembly:
                            </p>
            
                            <center>
                                <div class="table-cont">
                                    <table class="small-table">
                                        <tr>
                                            <td class="tcol-1"><code class="inline">L</code></td><td class="tcol-2">150</td>
                                        </tr>
                                        <tr>
                                            <td class="tcol-1"><code class="inline">N</code></td><td class="tcol-2">537507 * 2</td>
                                        </tr>
                                        <tr>
                                            <td class="tcol-1"><code class="inline">G</code></td><td class="tcol-2">32000000</td>
                                        </tr>
                                    </table>
                                </div>
                            </center>
            
                            <p><span class="math display" id="eqn">\[ Coverage = \frac{150 * 537507 * 2}{32000000} = 5.04 \]</span></p>
            
                            <p>So we estimate our coverage at <b>5.04</b>, meaning that we would expect every position to be covered by 5 reads.</p>
            
                            <p>Let's see if this matches up to the observed read depth. First, lets use <code class="inline">samtools depth</code> to get the number
                                of reads that map to each position.
                            </p>
            
                            <center><pre class="cmd"><code>samtools depth dpse2-to-dmel3R.mkdup.bam > dpse2-to-dmel3R-depth.tab</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">samtools</td><td class="tcol-2">Call the <code class="inline">samtools</code> program.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">depth</td><td class="tcol-2">Use the <code class="inline">depth</code> sub-program within <code class="inline">samtools</code>.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R.mkdup.bam</td><td class="tcol-2">The input BAM/SAM file.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">&gt;</td><td class="tcol-2">This redirects the output that would have been printed to the screen to a file instead.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R-depth.tab</td><td class="tcol-2">The desired output file</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>The <code class="inline">dpse2-to-dmel3R-depth.tab</code> file simply contains the position in the reference genome and the number of reads mapped to this position.
                                These files can get very large for large genomes, so its generally not adviseable to plot read depth per-base (as we're about to do). Instead, for larger genomes,
                                windows of 10kb to 100kb can be averaged and analyzed together. Another good program for calculating per-base read depth with some clever space-saving output
                                options is <a href="https://github.com/brentp/mosdepth" target="_blank">mosdepth</a>.
                            </p>
            
                            <p>But since we're looking at a small genome, let's just analyze this file! One of the pre-baked scripts provided is an R script that takes a 
                                <code class="inline">samtools depth</code> file as input and calculates averge read depth and generates some figures. Let's run it on the file
                                we just created:
                            </p>
            
                            <center><pre class="cmd"><code>Rscript ../scripts/depth_plot.r dpse2-to-dmel3R-depth.tab</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">Rscript</td><td class="tcol-2">Invokes R to run the script.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">../scripts/depth_plot.r</td><td class="tcol-2">The path to the script.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">dpse2-to-dmel3R-depth.tab</td><td class="tcol-2">The input file.</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>This script will take a couple of minutes to run, but should print the average read depth to the screen a histogram of read depth:
                                <code class="inline">depth-hist.png</code>, which you can look at by clicking on it in the file browser if you'd like. I see that average read depth is about 4:
                            </p>
            
                            <pre class="text"><code># Avg. read depth: 3.948774</code></pre>
            
                            <p>This is lower than expected based on our calculation with the Lander/Waterman equation. Why? One cause would be many reads having sequencing errors leading to 
                                unmapped reads. This doesn't seem to be the case for our data, based on the <a href="reads.html#fastqc-bq">FastQC results we obtained earlier</a>. Another cause could be that the reads that 
                                contain the most variation between the two species were not able to be mapped. This is referred to as <b>reference bias</b>: reads that contain too
                                much variation cannot be mapped, so the final mapped reads end up looking more like the reference genome than they actually are. <b>Iterative mapping</b> 
                                attempts to address this issue.
                            </p>                    

                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

        </div>
    </div>

    <div class="row" id="btm-nav">
        <div class="col-3-24" id="nav-bnt-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="assembly.html">&lt;&nbsp;Previous</a>    
            </div>
        </div>
        <div class="col-6-24" id="nav-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="iterative-mapping.html">Next&nbsp;&gt;</a>
            </div>
        </div>
        <div class="col-3-24" id="nav-btn-margin"></div>
    </div>

    <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
        (function () {
            var script = document.createElement("script");
            script.type = "text/javascript";
            script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
            document.getElementsByTagName("head")[0].appendChild(script);
        })();
    </script>

    <div class="row" id="footer">
		<div class="col-1">
			<div id="footer_text">
				<center>Site designed and maintained by <a href="https://gwct.github.io" target="_blank">Gregg Thomas</a> | Page built: 01/21/2021 21:47:37 MST</center>
			</div>
		</div>
	</div>
</body>
