<a class="internal-link" name="fasta"></a>
<div class="row" id="header">Sequences & Reads</div>

    <div class="row" id="body-row">
        <div class="col-3-24" id="side-nav-cont">
            <div id="side-nav">
                <span id="side-header">Page contents</span>
                <ul>
                    <li><a href="lab01-2.html#fasta">FASTA format</a></li>
                    <li><a href="lab01-2.html#fastq">FASTQ format</a></li>
                    <li><a href="lab01-2.html#phred">Phred Quality scores</a></li>
                    <li><a href="lab01-2.html#act1">ACTIVITY 1</a></li>
                    <li><a href="lab01-2.html#fastqc">Assessing reads with Fastqc</a></li>
                </ul>
            </div>
        </div>

        <div class="col-21-24" id="main-cont">
            <div class="row" id="top-row-cont">
                <div class="col-24-24" id="top-row"></div>
            </div>

            <div class="row" id="top-row-cont">
                <div class="col-24-24" id="top-row"></div>
            </div>

            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">FASTA format</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">
        
                            <p>One common format used to store biological sequence data is the FASTA format. Let's take a look at one of the reference sequences stored in FASTA format with the following command:</p>
        
                            <center><pre class="cmd"><code>less -S data/dmel-3R-reference/Drosophila_melanogaster.BDGP6.28.dna.chromosome.3R.fa</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">less</td><td class="tcol-2">A Linux text file viewer 
                                            (use <code class="inline">&lt;up arrow&gt;</code> and <code class="inline">&lt;down arrow&gt;</code> to scroll up and down by one line 
                                            or <code class="inline">&lt;space&gt;</code> and <code class="inline">b</code> 
                                            to scroll up and down by one page, resepectively. Type <code class="inline">q</code> to exit)</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-S</td><td class="tcol-2">Turn off line-wrapping within <code class="inline">less</code> 
                                            (use the <code class="inline">&lt;left arrow&gt;</code> and <code class="inline">&lt;right arrow&gt;</code> to scroll left and right).</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">data/dmel-3R-reference/Drosophila_melanogaster.BDGP6.28.dna.chromosome.3R.fa</td><td class="tcol-2">The path to the file you want to view</td>
                                    </tr>
                                </table>
                            </div>
            
                            <p>Here I've clipped the top of this output:</p>
            
                            <pre class="text"><code>>3R dna:chromosome chromosome:BDGP6.28:3R:1:32079331:1 REF
ACGGGACCGAGTATAGTACCAGTACGCGACCAGTACGGGAGCAGTACGGAACCAATACGG
GTCCAGTACGGGTCCAGTACGGGTCCAGTACGGGTCCAGTACGGGACCAGTACGGGACCG</code></pre>
            
                            <p>We can see that the first line begins with the greater than symbol (<code class="inline">></code>). This indicates that it is a header line for the following sequence.
                                Header lines can contain whatever information the file creator thought would be important to annotate the sequence. In most cases, it will have a name
                                for the sequence and some information about its location in the genome.</p>
            
                            <p>After the header line comes the sequence. Sequences can be split up to multiple lines, but always form a single string that refers back to the previous header. 
                                There can also be multiple headers/sequences in a single FASTA file.</p>
                                
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>
        
            <a class="internal-link" name="fastq"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">FASTQ format</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">
        
                            <p>Reads require a different format because they are need to store another important piece of information regarding sequence data: per-base quality scores.
                                To accomplish this, the FASTQ format is used. We can see an example of FASTQ format with the reads provided:</p>
            
                            <center><pre class="cmd"><code>zless -S data/dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">zless</td><td class="tcol-2">A Linux compressed text file viewer 
                                            (use <code class="inline">&lt;up arrow&gt;</code> and <code class="inline">&lt;down arrow&gt;</code> to scroll up and down by one line 
                                            or <code class="inline">&lt;space&gt;</code> and <code class="inline">b</code> 
                                            to scroll up and down by one page, resepectively. Type <code class="inline">q</code> to exit)</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-S</td><td class="tcol-2">Turn off line-wrapping within <code class="inline">zless</code> 
                                            (use the <code class="inline">&lt;left arrow&gt;</code> and <code class="inline">&lt;right arrow&gt;</code> to scroll left and right).</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">data/dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz</td><td class="tcol-2">The path to the file you want to view</td>
                                    </tr>
                                </table>
                            </div>
            
            
                            <p>Here is what the first few lines of that file should look like:</p>
                            <pre class="text"><code>@SRR11230564.32890235/1
TGATAATAAAACAACTTATTGCCCCCCTCTGGCAATACACTCTCCCGGTGTGGGGGGCCCTGACATGTAAAACGCAAATAAGAAGAATCCAGGGACAGCAGAACAAAATAATCAGAAATGCACTACTAGATCGGAAGAGCACACGTCTGAA
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ
@SRR11230564.32288433/1
ATTGTGTGTATGTTGTATTTTTCCCTAAGCACTTCATTTCTTGTGTAGCGTCCAGCTAGTAGTGCATTTCTGATTATTTTGTTCTGCTGTCCCTGGATTCTTCTTATTTGCGTTTTACATGTCAGGGCCCCCCACACCGGGAGATCGGAAG
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJ&lt;JJJJJAJJJJJJFJF</code></pre>
    
                            <p>So as you can see there's a lot more information here. In FASTQ, lines that start with the <code class="inline">@</code> character are the header lines and
                                can contain any amount of information about the sequence, including its ID, information about the sequencing machine that generated it, and whether
                                it is one of a pair of reads. In this case, since these reads were downloaded from the SRA, only an ID and the pair number (<code class="inline">/1</code>)
                                are included. In the snippet above there are two lines starting with <code class="inline">@</code>, meaning we are looking at information for two reads.</p>
            
                            <p>Following the header line are lines that contain sequences corresponding to that header, just like in FASTA formatted files. However, after this
                                is a line that starts with <code class="inline">+</code>. This is a linker line between the sequence and the quality scores. It can be left blank with just the
                                <code class="inline">+</code> sign, or can contain header information as well.
                            </p>
            
                            <p>After the linker line are the <b>ASCII encoded Phred-scaled quality scores.</b></p>                        
        
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="phred"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Phred quality scores</div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">
            
                            <p>Sequencing machines read sequences as signals such as fluorescence or changes in pH or charge. Many sequencing machines can measure the intensity
                                of these signals and other factors at run-time to assign the probabilty that the base called at each position is an error. These error probabilities
                                are transformed into quality scores as follows:</p>
            
                            <p><span class="math display" id="eqn">\[ Q = -10\log P(error)\]</span></p>
            
                            <p>Scores resulting from this transformation are known as Phred scores because they were first used by the base-calling software Phred and usually range from 0 to 40.</p>
            
                            <p>To encode these scores into FASTQ files, they are converted to an ASCII character. Briefly, starting with the number 33, each character on an English keyboard
                                has a corresponding integer code (called the decimal value or DEC). To convert Phred scores to ASCII characters, we first add 33 to the Phred score and look up that
                                value in the ASCII table.</p>
            
                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="res-img" src="img/ascii-table.png">
                                    <center><span class="fig-caption">Figure 2.1: ASCII table.</span></center>
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>
            
                            <p>Converting from ASCII characters back to error probabilities is just the reverse of this process. For example, the first ASCII character in the quality strin
                                of the first read above is <code class="inline">A</code>. Looking this up in the table, <code class="inline">A</code> has a DEC value of 65. The corresponding Phred score would then be
                                65 - 33 = 32. And then we convert back to error probability with the reverse of the above equation:
                            </p>
            
                            <p><span class="math display" id="eqn">\[ P(error) = 10^{-\frac{Q}{10}} \]</span></p>
            
                            <p>Using this equation, a score of 32 means the probability that this base is called erroneously is 0.00063</p>

                            <a class="internal-link" name="act1"></a>
                            <h2>ACTIVITY: Converting between quality score encodings and error probabilities</h2>

                            <p>
                                Using the equations and table above, fill in the following table and answer the questions below. Use any type of calculating/conversion tool you need.
                            </p>

                            <div class="table-cont">
                                <table class="main-table">
                                    <thead><th class="tcol-1">ASCII character</th><th class="tcol-2">Phred score</th><th class="tcol-2">Error probability</th></thead>
                                    <tr>
                                        <td class="tcol-1"> </td>
                                        <td class="tcol-2">40</td>
                                        <td class="tcol-2"> </td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">!</td>
                                        <td class="tcol-2"> </td>
                                        <td class="tcol-2"> </td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1"> </td>
                                        <td class="tcol-2"> </td>
                                        <td class="tcol-2">0.000316</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">5</td>
                                        <td class="tcol-2"> </td>
                                        <td class="tcol-2"> </td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1"> </td>
                                        <td class="tcol-2"> </td>
                                        <td class="tcol-2">0.000126</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">?</td>
                                        <td class="tcol-2"> </td>
                                        <td class="tcol-2"> </td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1"> </td>
                                        <td class="tcol-2">12</td>
                                        <td class="tcol-2"> </td>
                                    </tr>
                                </table>
                            </div>

                            <h3>Questions</h3>

                            <ol>
                                <li>
                                    What Phred score is associated with a base call <b>accuracy</b> (1 - P(error)) of 99.99%? 99.9%?
                                </li>

                                <li>
                                    What do you think would be a good cut-off for quality filtering? In other words, at what quality score would you think it reasonable
                                    to remove all bases that score at or below that threshold? Remember, the human genome has roughly 3 BILLION bases.
                                </li>
                            </ol>
            
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

            <a class="internal-link" name="fastqc"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Assessing read quality with <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a></div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">
        
                            <p>One of the first things one should do after receiving the sequence data (reads) is to assess their quality. This means acquiring some basic
                                statistics, such as number of reads, average read length, average quality, etc.
                            </p>
            
                            <p>Fortunately, since this is such a common task, there are programs out there that help. One of the most widely used is <code class="inline">fastqc</code>.
                                This program takes as input a set of reads in <b>FASTQ</b> format and generates an HTML file with several plots that can be viewed within a web browser.
                            </p>
            
                            <p>Let's try to run <code class="inline">fastqc</code> on some of our short read (Illumina) data for <em>D. pseudoobscura</em> chromosome 2.</p>
            
                            <center><pre class="cmd"><code>fastqc data/dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz data/dpse-chr2-reads/illumina/pse-chr2_2.fastq.gz -o fastqc-output/</code></pre></center>
            
                            <div class="table-cont">
                                <table class="cmd-table">
                                    <thead><th class="tcol-1">Command line parameter</th><th class="tcol-2">Description</th></thead>
                                    <tr>
                                        <td class="tcol-1">fastqc</td><td class="tcol-2">Invoke the <code class="inline">fastqc</code> program.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">data/dpse-chr2-reads/illumina/pse-chr2_1.fastq.gz</td><td class="tcol-2">The first input read file.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">data/dpse-chr2-reads/illumina/pse-chr2_2.fastq.gz</td><td class="tcol-2">The second input read file.</td>
                                    </tr>
                                    <tr>
                                        <td class="tcol-1">-o fastqc-output/</td><td class="tcol-2">The <code class="inline">-o</code> option specifies the output directory.</td>
                                    </tr>
                                </table>
                            </div>

                            <p>This should take less than a minute to run, during which you'll see some progress reported on the screen.</p>
            
                            <p>Unlike most of our other commands today, this won't output a text file that can be viewed easily from the terminal. Rather the HTML output should be viewed
                                in the browser. To do that, navigate to that folder on your computer using the OS's GUI. Once you're there you should see 4 files, 
                                two <code class="inline">.html</code> files and two <code class="inline">.zip</code> files. One file per type per input file. We can 
                                ignore the <code class="inline">.zip</code> for now. They contain the raw output if we ever want to look at that ourselves. But FastQC 
                                helpfully summarizes it with the HTML files. Double click on the file <code class="inline">pse-chr2_1_fastqc.html</code> 
                                to open the file in a new browswer tab.
                            </p>
            
                            <!-- <div id="msg_cont">
                                <div id="msg">
                                    <div id="msg_banner">Tip - Unblock pop-ups</div>
                                    <div id="msg_text">
                                        <p>
                                            The first time you select <code class="inline">View in Web Browser</code> you may get a warning that your browser does not allow pop-ups. There
                                            should also be an option to allow pop-ups from this site. Click on that option and the file should open in a new tab. If not, click on the file and
                                            select <code class="inline">View in Web Browser</code> again to open.
                                        </p>
                                        <p></p>
                                    </div>
                                </div>
                            </div> -->

                            <p></p>

                            <!-- <div class="row" id="img-row">
                                <div class="col-24-24" id="img-col">
                                    <img id="res-img" src="img/fastqc-screenshot.png">
                                    <center><span class="fig-caption">Figure 2.2: A screenshot of the RStudio interface showing how to view an HTML file.</span></center>
                                </div>
                            </div>                             -->

                            <p>Once you have it open, the FastQC output should display a number of plots and summary statistics which can be scrolled through or navigated to with the links on the left.</p>
            
                            <div class="row" id="img-row">
                                <div class="col-7-24" id="margin"></div>
                                <div class="col-10-24" id="img-col">
                                    <img id="res-img" src="img/fastqc-1.png">
                                    <center><span class="fig-caption">Figure 2.3: Basic FastQC statistics from <code class="inline">pse-chr2_1.fastq.gz</code>.</span></center>
                                </div>
                                <div class="col-7-24" id="margin"></div>
                            </div>
            
                            <p>This tells us we've sequenced <b>537507</b> reads that range from 30-151bp long (though the vast majority are 150bp). It also tells us that none of the reads
                                are considered low quality by FastQC. That's good! But let's take a look at the quality score distribution anyway:
                            </p>
            
                            <a name="fastqc-bq"></a>	
                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="res-img" src="img/fastqc-2.png">
                                    <center><span class="fig-caption">Figure 2.4: Distribution of <b>Phred quality scores</b> based on position in read in <code class="inline">pse-chr2_1.fastq.gz</code>.</span></center>
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>
                            
                            <p>This shows that for most positions in most reads we have the highest quality score possible (40), with a bit of a drop-off towards the end of reads. This
                                drop-off towards the end is usually caused by <b>de-phasing</b>. De-phasing happens because of the random nature of sequencing by synthesis. At any given spot
                                on a flow-cell there may be thousands of cloned PCR fragments to be sequenced. During each cycle of the machine, nucleotides are washed over these spots and
                                incorporated into the sequence -- but not in 100% of the fragments at a given spot. Just by chance, some sequences may not be extended in a given cycle, adding
                                noise to the signal, which accumulates towards the ends of reads.
                            </p>
            
                            <p>Despite this, I would still say this is high-quality data. Just to see, here is an example of poorer quality data from the FastQC homepage:</p>
            
                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="res-img" src="img/fastqc-3.png">
                                    <center><span class="fig-caption">Figure 2.5: An example of poor quality reads. 
                                        <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">Source</a>.</span></center>
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>
            
                            <p>The only red flag in our reads is adapter content. Recall that adapters are the short sequences of nucleotides that allow the cloned DNA fragments to
                                affix to the flowcell. Needless to say we wouldn't want to incorporate these sequences in our assemblies. Fortunately, most assemblers and read mappers
                                trim these off automatically.    
                            </p>
            
                            <p>There are several pre-processing tasks you could perform to filter your reads, including:</p>
            
                            <ol>
                                <li>Trimming adapters</li>
                                <li>Trimming low quality positions</li>
                                <li>Decontamination</li>
                                <li>Removal of duplicate reads</li>
                            </ol>
            
                            <p>But in general, I allow the assembly and mapping programs to handle these tasks. The one exception being the marking of duplicate reads after read
                                mapping, which is something we'll do in the <a href="mapping.html">Read mapping</a> section. But for now, let's move on to
                                <a href="assembly.html">Assembly</a>...
                            </p>
                                                
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>

        </div>
    </div>

    <div class="row" id="btm-nav">
        <div class="col-3-24" id="nav-bnt-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="lab01-1.html">&lt;&nbsp;Previous</a>    
            </div>
        </div>
        <div class="col-6-24" id="nav-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="lab01-3.html">Next&nbsp;&gt;</a>
            </div>
        </div>
        <div class="col-3-24" id="nav-btn-margin"></div>
    </div>

    <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
        (function () {
            var script = document.createElement("script");
            script.type = "text/javascript";
            script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
            document.getElementsByTagName("head")[0].appendChild(script);
        })();
    </script>