<a class="internal-link" name="intro"></a>
<div class="row" id="header">Sequences & Reads</div>

    <div class="row" id="body-row">
        <div class="col-3-24" id="side-nav-cont">
            <div id="side-nav">
                <span id="side-header">Page contents</span>
                <ul>
                    <li><a href="lab01-3.html#part1">Part 1</a></li>
                    <li><a href="lab01-3.html#part2">Part 2</a></li>
                    <li><a href="lab01-3.html#part3">Part 3</a></li>
                </ul>
            </div>
        </div>

        <div class="col-21-24" id="main-cont">
            <div class="row" id="top-row-cont">
                <div class="col-24-24" id="top-row"></div>
            </div>

            <div class="row" id="top-row-cont">
                <div class="col-24-24" id="top-row"></div>
            </div>

            <a class="internal-link" name="assign"></a>
            <div class="row" id="section-header-cont">
                <div class="col-24-24" id="section-header-row">
                    <div id="section-header">Assignment: Assessing base quality</a></div>
                </div>
            </div>
            <div class="row" id="section-cont">
                <div class="col-24-24" id="section-col">
                    <div class="row" id="section-row">
                        <div class="col-2-24" id="inner-margin"></div>
                        <div class="col-20-24" id="section-content">
        
                            <a class="internal-link" name="part1"></a>
                            <h2>Part 1 - Write a script to count base quality scores</h2>

                            <p>
                                In groups of 3, write a script that takes as input a FASTQ file and outputs a table in TSV format that counts the number of bases with each Phred score.
                            </p>

                            <h3>Input and expected outputs</h3>
                            <p>
                                The output file should have 3 columns: Phred score, Probability of error, # of bases with score.
                            </p>

                            <p>
                                The output file should have 42 rows (excluding the header): 1 row for each Phred score from 0 to 41.
                            </p>

                            <p>
                                I will be using the <em>D. pseudoobscura</em> reads from lab to assess your script, so you should use those as your input data, but your code should
                                work on ANY FASTQ file!
                            </p>

                            <h3>Collaborative coding</h3>

                            <p>
                                Feel free to split up the work as you see fit, but collaborative coding is a team effort so everyone must contribute! Be sure to leave comments on chunks
                                of code/functions that indicate who wrote them.
                            </p>

                            <p>
                                A possible way to divide the work:
                            </p>

                            <ul>
                                <li>One group member writes the "scaffolding" code. This includes parsing input options (including error checking),
                                    data structure initialization, and writing output files.</li>
                                <li>One group member writes a function that reads the FASTQ file line-by-line, passing each line to a counting function.</li>
                                <li>One group member writes a function that takes a string of quality scores and counts the occurrence of each score, 
                                    adding that count to the overall counts.</li>
                            </ul>

                            <h3>Considerations</h3>

                            <p>
                                What data structure will be best to use when tracking counts of each score?
                            </p>

                            <h3>Example outputs</h3>

                            <p>
                                The table should be formatted as follows:
                            </p>

                            <div class="table-cont">
                                <table class="main-table">
                                    <thead><th class="tcol-1">Phred score</th><!-- <th class="tcol-1">ASCII character</th> --><th class="tcol-1">Error probability</th><th class="tcol-1">Number of bases with score</th></thead>
                                    <tr>
                                        <td class="tcol-1">0</td>
                                        <!-- <td class="tcol-1">!</td> -->
                                        <td class="tcol-1">1.0</td>
                                        <td class="tcol-1">0</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">1</td>
                                        <!-- <td class="tcol-1">"</td> -->
                                        <td class="tcol-1">0.7943282347242815</td>
                                        <td class="tcol-1">0</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">2</td>
                                        <!-- <td class="tcol-1">#</td> -->
                                        <td class="tcol-1">0.6309573444801932</td>
                                        <td class="tcol-1">10654</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">.</br>.</br>.</br></td>
                                        <!-- <td class="tcol-1"> </td> -->
                                        <td class="tcol-1"> </td>
                                        <td class="tcol-1"> </td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">40</td>
                                        <!-- <td class="tcol-1">I</td> -->
                                        <td class="tcol-1">0.0001</td>
                                        <td class="tcol-1">0</td>
                                    </tr>

                                    <tr>
                                        <td class="tcol-1">41</td>
                                        <!-- <td class="tcol-1">J</td> -->
                                        <td class="tcol-1">7.943282347242822e-05</td>
                                        <td class="tcol-1">66987678</td>
                                    </tr>
                                </table>
                            </div>

                            <h4><a href="genomics/lab01/scripts/fastq-count.txt" download>Example Output</a></h4>

                            <a class="internal-link" name="part2"></a>
                            <h2>Part 2 - Visualizing the results</h2>

                            <p>
                                Write an R script that plots the distribution of scores. Since you've already binned the data, this would be a simple bar graph.
                                Style the figure however you'd like.
                            </p>

                            <div class="row" id="img-row">
                                <div class="col-6-24" id="margin"></div>
                                <div class="col-12-24" id="img-col">
                                    <img id="main-img" src="genomics/lab01/scripts/score-counts.png">
                                    <center><span class="fig-caption">The number of bases with each Phred score in the <em>D. pseudooboscure</em> data.</span></center> 
                                </div>
                                <div class="col-6-24" id="margin"></div>
                            </div>      
                
                            <a class="internal-link" name="part3"></a>
                            <h2>Part 3 - Written questions</h2>

                            <ol>
                                <li>
                                    Is this high quality data?
                                </li>

                                <li>
                                    Using the cutoff you decided on in lab, how many bases would have been excluded from the current dataset?
                                </li>

                                <li>
                                    Given a genome that has 3 billion bases assume you have a set of reads that covers each base exactly once AND that they all
                                    have the quality score you chose for your cutoff. How many bases are actually errors?
                                </li>
                            </ol>

                            <div id="sep_div"></div>
                        </div>
                        <div class="col-2-24" id="inner-margin"></div>
                    </div>
                </div>
            </div>



        </div>
    </div>

    <div class="row" id="btm-nav">
        <div class="col-3-24" id="nav-bnt-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <div class="nav-btn">
                <a href="lab01-2.html">&lt;&nbsp;Previous</a>    
            </div>
        </div>
        <div class="col-6-24" id="nav-margin"></div>
        <div class="col-6-24" id="nav-btn-cont">
            <!-- <div class="nav-btn">
                <a href="assembly.html">Next&nbsp;&gt;</a>
            </div> -->
        </div>
        <div class="col-3-24" id="nav-btn-margin"></div>
    </div>

    <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
        (function () {
            var script = document.createElement("script");
            script.type = "text/javascript";
            script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
            document.getElementsByTagName("head")[0].appendChild(script);
        })();
    </script>